crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -308,298 +373,261 @@ impl AppHandlerExecute for Ics20Transfer {
+    // the packet was not commited on the counter party chain but returned as-is.
+    let receiver = parse_sender(&state, &packet_data.sender)
         .await
-        .context("failed to parse ibc packet sender as the refund target address")?;
-    execute_deposit(
-        state,
-        bridge_address,
-        &denom,
+        .with_context(|| {
+            format!(
+                "failed parsing packet.sender `{}` as the return address",
+                packet_data.sender
+            )
+        })?;
+
+    let asset = parse_asset(&state, &packet_data.denom)
+        .await
+        .with_context(|| format!("failed parsing packet.asset `{}`", packet_data.denom))?;
+
+    // Refunding a rollup is the same as refunding an address on sequencer (which would
+    // be the bridge account associated with the rollup) plus emitting a deposit.
+    if let Some(memo) = does_failed_transfer_come_from_rollup(&packet_data) {
+        emit_deposit(
+            &mut state,
+            receiver,
+            memo.rollup_return_address,
+            &asset,
+            amount,
+        )
+        .await
+        .context("failed to emit deposit for refunding tokens to rollup")?;
+    }
+    refund_tokens_to_sequencer_address(
+        &mut state,
+        receiver,
+        asset,
         amount,
-        bridge_address.to_string(),
+        &packet.port_on_a,
+        &packet.chan_on_a,
     )
     .await
-    .context("failed to emit deposit")?;
+    .context("failed to refund a sequencer address")?;
 
-    state
-        .increase_balance(bridge_address, denom, amount)
-        .await
-        .wrap_err("failed to update bridge account account balance")?;
+    Ok(())
+}
+
+/// A failed transfer is said to originate on a rollup if its memo field can be
+/// parsed as a `[Ics20WithdrawalFromRollup]`.
+fn does_failed_transfer_come_from_rollup(
+    packet_data: &FungibleTokenPacketData,
+) -> Option<Ics20WithdrawalFromRollup> {
+    serde_json::from_str::<Ics20WithdrawalFromRollup>(&packet_data.memo).ok()
+}
 
+#[instrument(skip_all, fields(%recipient, %asset, amount), err)]
+async fn refund_tokens_to_sequencer_address<S: StateWrite>(
+    mut state: S,
+    recipient: Address,
+    asset: denom::TracePrefixed,
+    amount: u128,
+    source_port: &PortId,
+    source_channel: &ChannelId,
+) -> Result<()> {
+    if is_refund_source_zone(&asset, source_port, source_channel) {
+        state
+            .decrease_ibc_channel_balance(source_channel, &asset, amount)
+            .await
+            .context("failed to withdraw refund amount from escrow account")?;
+        state
+            .increase_balance(recipient, &asset, amount)
+            .await
+            .context("failed to increase user account balance")?;
+    } else {
+        state
+            .increase_balance(recipient, asset, amount)
+            .await
+            .wrap_err("failed to update user account balance in execute_ics20_transfer")?;
+    }
```

Comment:
```suggestion
    if is_refund_source_zone(&asset, source_port, source_channel) {
        state
            .decrease_ibc_channel_balance(source_channel, &asset, amount)
            .await
            .context("failed to withdraw refund amount from escrow account")?;
    } 
    
   state
          .increase_balance(recipient, asset, amount)
          .await
          .wrap_err("failed to update user account balance when refunding")?;
 
```

Reply:
Nice!

Reply:
I did the same thing in `receive_tokens`! 🥳